<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.550">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Pankaj Chejara">
<meta name="dcterms.date" content="2024-11-19">

<title>bioinformatics - Metagenomic Analysis of Gut Microbiome for their Association with Colorectal Cancer</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="fullcontent">

<div id="quarto-search-results"></div>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Metagenomic Analysis of Gut Microbiome for their Association with Colorectal Cancer</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">python</div>
                <div class="quarto-category">bioinformatics</div>
                <div class="quarto-category">metagenomics</div>
                <div class="quarto-category">colorectal cancer</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Pankaj Chejara </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">November 19, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>This post delves into the field of bioinformatics to illustrate the potential of metagenomic data to predict colorectal cancer. Metagenomics data are produced by high-throughput sequencing machines and typically require a substantial amount of compute and storage resources.</p>
<p>In this post, we will utilize <strong>processed metagenomics data from different cohorts on colorectal cancer</strong>. The dataset (relative abundance) is available <a href="http://segatalab.cibio.unitn.it/data/Thomas_et_al.html">here</a>.</p>
<p>The <strong>relative abundance dataset</strong> contains samples from datasets from previous research studies as well as their own datasets. The below table provides a summary of cohorts and their information.</p>
<table class="table">
<colgroup>
<col style="width: 25%">
<col style="width: 35%">
<col style="width: 22%">
<col style="width: 16%">
</colgroup>
<thead>
<tr class="header">
<th>Citation</th>
<th>Groups</th>
<th>Country</th>
<th>Accession no.</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Zeller et al., 2014</td>
<td>Control(61) Adenoma(42) CRC(53)</td>
<td>France</td>
<td>PRJEB6070</td>
</tr>
<tr class="even">
<td>Yu et al., 2015</td>
<td>Control(54) CRC(74)</td>
<td>China</td>
<td>ERP008729</td>
</tr>
<tr class="odd">
<td>Feng et al., 2015</td>
<td>Control(61) Adenoma(47) CRC(46)</td>
<td>Austria</td>
<td>ERP005534</td>
</tr>
<tr class="even">
<td>Vogtmann et al., 2016</td>
<td>Control(52) CRC(52)</td>
<td>USA</td>
<td>PRJEB6070</td>
</tr>
<tr class="odd">
<td>Hannigan et al., 2018</td>
<td>Control(28) Adenoma(27) CRC(29)</td>
<td>USA (54) Canada(28)</td>
<td>PRJNA389927</td>
</tr>
<tr class="even">
<td>Thomas et al., 2019</td>
<td>Control(24) Adenoma(27) CRC(29)</td>
<td>Italy</td>
<td>SRP136711</td>
</tr>
<tr class="odd">
<td>Thomas et al., 2019a</td>
<td>Control(28) CRC(32)</td>
<td>Italy</td>
<td>SRP136711</td>
</tr>
</tbody>
</table>
<p><br></p>
<div class="callout callout-style-simple callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Downloading dataset using accession number
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Here, we will see a script to automatically download the raw sequence paired data using project accession number.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Enter the project identifier (hint: it will be starting from PR...)"</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">read</span> <span class="va">project</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Preparing for the download"</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># check if project directory exists</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">[</span> <span class="ot">-d</span> <span class="st">"</span><span class="va">$project</span><span class="st">"</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">"  </span><span class="va">$project</span><span class="st"> already exists. cd to </span><span class="va">$project</span><span class="st">"</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">"  creating a new directory </span><span class="va">$project</span><span class="st">"</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mkdir</span> <span class="va">$project</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="cf">fi</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># change to project directory</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$project</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"  changing to </span><span class="va">$project</span><span class="st"> directory"</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co"># check if raw_data directory exists</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">[</span> <span class="ot">-d</span> <span class="st">"raw_data"</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">"  raw_data already exists. cd to </span><span class="va">$raw_data</span><span class="st">"</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">"  creating a new directory raw_data"</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mkdir</span> raw_data</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="cf">fi</span></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a><span class="co"># run search query and store metadata</span></span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a><span class="ex">esearch</span> <span class="at">-db</span> sra <span class="at">-query</span> <span class="va">$project</span>  <span class="kw">|</span> <span class="ex">efetch</span> <span class="at">-format</span> runinfo <span class="op">&gt;</span> <span class="va">${project}</span>_metadata.txt</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a><span class="co"># fetch accessions numbers from metada file into a text file</span></span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="va">${project}</span>_metadata.txt <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f</span> 1 <span class="at">-d</span> <span class="st">','</span> <span class="kw">|</span><span class="fu">grep</span> <span class="st">"SRR"</span> <span class="op">&gt;</span> <span class="va">${project}</span>_sra_id.txt</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a><span class="co"># change directory to raw_data</span></span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> raw_data</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"  changing to raw_data directory"</span></span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a> <span class="co"># download sra file, fastq file and zip fastq file into gz</span></span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> ../<span class="va">${project}</span>_sra_id.txt <span class="kw">|</span> <span class="cf">while</span> <span class="bu">read</span> <span class="va">sra_id</span><span class="kw">;</span> <span class="cf">do</span> <span class="ex">prefetch</span> <span class="va">$sra_id</span><span class="kw">;</span> <span class="ex">fasterq-dump</span> <span class="va">$sra_id</span><span class="kw">;</span> <span class="fu">gzip</span> <span class="va">${sra_id}</span><span class="pp">*</span>.fastq<span class="kw">;</span><span class="cf">done</span></span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Sequence data has been downloaded successfully"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<section id="accessing-abundance-data" class="level2">
<h2 class="anchored" data-anchor-id="accessing-abundance-data">Accessing abundance data</h2>
<p>Here, we will utilize publicly available processed data of all aforementioned publications. The <strong>dataset is publicly available</strong> and can be downloaded from <a href="http://segatalab.cibio.unitn.it/data/Thomas_et_al.html">here</a>.</p>
<p>Let’s load the dataset first and then explore it a bit.</p>
<section id="loading-the-dataset" class="level3">
<h3 class="anchored" data-anchor-id="loading-the-dataset">Loading the dataset</h3>
<div id="894f0d26-9b58-447b-b72d-4b240588a7c2" class="cell" data-execution_count="1">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> pd.read_csv(<span class="st">'Nine_CRC_cohorts_taxon_profiles.tsv'</span>,sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>,header<span class="op">=</span><span class="va">None</span>).T</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> data.rename(columns<span class="op">=</span>data.loc[<span class="dv">0</span>]).drop(<span class="dv">0</span>, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>data.head(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="1">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">dataset_name</th>
<th data-quarto-table-cell-role="th">sampleID</th>
<th data-quarto-table-cell-role="th">subjectID</th>
<th data-quarto-table-cell-role="th">body_site</th>
<th data-quarto-table-cell-role="th">study_condition</th>
<th data-quarto-table-cell-role="th">disease</th>
<th data-quarto-table-cell-role="th">age</th>
<th data-quarto-table-cell-role="th">age_category</th>
<th data-quarto-table-cell-role="th">gender</th>
<th data-quarto-table-cell-role="th">country</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Circoviridae|g__Gyrovirus|s__Chicken_anemia_virus</th>
<th data-quarto-table-cell-role="th">k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Geminiviridae|g__Mastrevirus|s__Wheat_dwarf_virus</th>
<th data-quarto-table-cell-role="th">k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Hepadnaviridae|g__Orthohepadnavirus|s__Orthohepadnavirus_unclassified</th>
<th data-quarto-table-cell-role="th">k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Inoviridae|g__Inovirus|s__Enterobacteria_phage_I2_2</th>
<th data-quarto-table-cell-role="th">k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Inoviridae|g__Inovirus|s__Enterobacteria_phage_If1</th>
<th data-quarto-table-cell-role="th">k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Polyomaviridae|g__Polyomavirus|s__JC_polyomavirus</th>
<th data-quarto-table-cell-role="th">k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Retroviridae|g__Gammaretrovirus|s__Porcine_type_C_oncovirus</th>
<th data-quarto-table-cell-role="th">k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Retroviridae|g__Retroviridae_noname|s__Avian_endogenous_retrovirus_EAV_HP</th>
<th data-quarto-table-cell-role="th">k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Viruses_noname|g__Viruses_noname|s__Enterobacteria_phage_HK544</th>
<th data-quarto-table-cell-role="th">k__Viruses|p__Viruses_noname|c__Viruses_noname|o__Viruses_noname|f__Viruses_noname|g__Viruses_noname|s__Porcine_stool_associated_circular_virus_3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">1</td>
<td>ZellerG_2014</td>
<td>CCIS27304052ST-3-0</td>
<td>FR-001</td>
<td>stool</td>
<td>control</td>
<td>none</td>
<td>52</td>
<td>adult</td>
<td>female</td>
<td>FRA</td>
<td>...</td>
<td>0.0</td>
<td>0.61117</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2</td>
<td>ZellerG_2014</td>
<td>CCIS13047523ST-4-0</td>
<td>FR-003</td>
<td>stool</td>
<td>control</td>
<td>none</td>
<td>70</td>
<td>senior</td>
<td>male</td>
<td>FRA</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0038</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>ZellerG_2014</td>
<td>CCIS15794887ST-4-0</td>
<td>FR-024</td>
<td>stool</td>
<td>control</td>
<td>none</td>
<td>37</td>
<td>adult</td>
<td>female</td>
<td>FRA</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4</td>
<td>ZellerG_2014</td>
<td>CCIS94603952ST-4-0</td>
<td>FR-025</td>
<td>stool</td>
<td>control</td>
<td>none</td>
<td>80</td>
<td>senior</td>
<td>female</td>
<td>FRA</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5</td>
<td>ZellerG_2014</td>
<td>CCIS74726977ST-3-0</td>
<td>FR-026</td>
<td>stool</td>
<td>adenoma</td>
<td>adenoma</td>
<td>66</td>
<td>senior</td>
<td>male</td>
<td>FRA</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>

<p>5 rows × 829 columns</p>
</div>
</div>
</div>
</div>
<p>The processed dataset contains <strong>metadata information</strong> such as <code>body_site</code>, <code>study_condition</code> along with <strong>abundance data of microbial species</strong>.</p>
<p>We will first start with <strong>Zeller et al., 2014</strong> dataset. For that we will perform filtering based on dataset name and <strong>extract Bacteria kingdom</strong> only.</p>
<div id="ae6fffd2-bf11-40f9-a110-b37b31d5046a" class="cell" data-execution_count="2">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># accessing Zeller et al., 2014 dataset</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>zeller_db <span class="op">=</span> data.loc[data[<span class="st">'dataset_name'</span>] <span class="op">==</span> <span class="st">'ZellerG_2014'</span>,:]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># fetching microbacterial organism information-related columns</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>bacteria_colnames <span class="op">=</span> [col <span class="cf">for</span> col <span class="kw">in</span> data.columns <span class="cf">if</span> <span class="st">'k__Bacteria'</span> <span class="kw">in</span> col]</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># metadata colnames</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>metadata_colnames <span class="op">=</span> [<span class="st">'dataset_name'</span>, <span class="st">'sampleID'</span>, <span class="st">'subjectID'</span>, <span class="st">'body_site'</span>, <span class="st">'study_condition'</span>,</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>       <span class="st">'disease'</span>, <span class="st">'age'</span>, <span class="st">'age_category'</span>, <span class="st">'gender'</span>, <span class="st">'country'</span>]</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co"># dataframe with bacterial species</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>zeller_b <span class="op">=</span> zeller_db[bacteria_colnames <span class="op">+</span> [<span class="st">'subjectID'</span>]]</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co">#zeller_b.columns = bacteria_colnames + ['subjectID']</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="co"># functions to process the data and prepare in a format supporting phyloseq analysis</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_sample_table(df):</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df[metadata_colnames]</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df.set_index(<span class="st">'subjectID'</span>)</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>    df.drop([<span class="st">'dataset_name'</span>,<span class="st">'sampleID'</span>],axis<span class="op">=</span><span class="dv">1</span>,inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_otu_table(df):</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df[bacteria_colnames <span class="op">+</span> [<span class="st">'subjectID'</span>]]</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>    df.columns <span class="op">=</span> [<span class="st">"OTU_</span><span class="sc">{}</span><span class="st">"</span>.<span class="bu">format</span>(<span class="bu">str</span>(ind)) <span class="cf">for</span> ind, col <span class="kw">in</span> <span class="bu">enumerate</span>(bacteria_colnames)] <span class="op">+</span> [<span class="st">'subjectID'</span>]</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>    new_header <span class="op">=</span> df[<span class="st">'subjectID'</span>]</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>    df_new <span class="op">=</span> df.drop([<span class="st">'subjectID'</span>],axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>    df_t <span class="op">=</span> df_new.T</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>    df_t.columns <span class="op">=</span> new_header</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df_t</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_taxa_table(list_of_otus):</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>    otu <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>    mapping <span class="op">=</span> {}</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>    taxa_cols <span class="op">=</span> [<span class="st">'kingdom'</span>,<span class="st">'phylum'</span>,<span class="st">'class'</span>,<span class="st">'order'</span>,<span class="st">'family'</span>,<span class="st">'genus'</span>,<span class="st">'species'</span>]</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> pd.DataFrame(columns<span class="op">=</span>taxa_cols)</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>    otu_mapping <span class="op">=</span> {}</span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>    otu_ids <span class="op">=</span> []</span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> ind, otu <span class="kw">in</span> <span class="bu">enumerate</span>(list_of_otus):</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> {}</span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> col <span class="kw">in</span> taxa_cols:</span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>            tmp[col] <span class="op">=</span> get_specific_label(otu, col)</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>        tmp_df <span class="op">=</span> pd.DataFrame([tmp])</span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> pd.concat([df,tmp_df], ignore_index<span class="op">=</span>[<span class="dv">0</span>])</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a>        otu_id <span class="op">=</span> <span class="st">"OTU_</span><span class="sc">{}</span><span class="st">"</span>.<span class="bu">format</span>(<span class="bu">str</span>(ind))</span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a>        otu_mapping[otu] <span class="op">=</span> otu_id</span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a>        otu_ids.append(otu)</span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a>    df[<span class="st">'OTU'</span>] <span class="op">=</span> [<span class="st">'OTU_</span><span class="sc">{}</span><span class="st">'</span>.<span class="bu">format</span>(<span class="bu">str</span>(ind)) <span class="cf">for</span> ind <span class="kw">in</span> df.index]</span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df.set_index(<span class="st">'OTU'</span>)</span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df</span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_specific_label(l, t):</span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a><span class="co">    This function parse the taxonomic assignment lable and fetch the specified information (e.g., kingdom, family)</span></span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a><span class="co">        l (str): string of taxonomy</span></span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a><span class="co">        t (str): string specifying the requested information (e.g., kingdom, family, genus, etc.)</span></span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb3-65"><a href="#cb3-65" aria-hidden="true" tabindex="-1"></a><span class="co">        str: requested heirarichal info </span></span>
<span id="cb3-66"><a href="#cb3-66" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb3-67"><a href="#cb3-67" aria-hidden="true" tabindex="-1"></a>    taxa_order <span class="op">=</span> {<span class="st">'kingdom'</span>:<span class="dv">0</span>,<span class="st">'phylum'</span>:<span class="dv">1</span>,<span class="st">'class'</span>:<span class="dv">2</span>,<span class="st">'order'</span>:<span class="dv">3</span>,<span class="st">'family'</span>:<span class="dv">4</span>,<span class="st">'genus'</span>:<span class="dv">5</span>,<span class="st">'species'</span>:<span class="dv">6</span>}</span>
<span id="cb3-68"><a href="#cb3-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-69"><a href="#cb3-69" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span>:</span>
<span id="cb3-70"><a href="#cb3-70" aria-hidden="true" tabindex="-1"></a>        specific_label <span class="op">=</span> l.split(<span class="st">'|'</span>)[taxa_order[t]]</span>
<span id="cb3-71"><a href="#cb3-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-72"><a href="#cb3-72" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> specific_label.strip().split(<span class="st">'__'</span>)[<span class="dv">1</span>]</span>
<span id="cb3-73"><a href="#cb3-73" aria-hidden="true" tabindex="-1"></a>    <span class="cf">except</span>:</span>
<span id="cb3-74"><a href="#cb3-74" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="st">'Unknown'</span></span>
<span id="cb3-75"><a href="#cb3-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-76"><a href="#cb3-76" aria-hidden="true" tabindex="-1"></a><span class="co"># Seperating data into different files</span></span>
<span id="cb3-77"><a href="#cb3-77" aria-hidden="true" tabindex="-1"></a>otu_table <span class="op">=</span> get_otu_table(zeller_db)</span>
<span id="cb3-78"><a href="#cb3-78" aria-hidden="true" tabindex="-1"></a>taxa_table <span class="op">=</span> get_taxa_table(bacteria_colnames)</span>
<span id="cb3-79"><a href="#cb3-79" aria-hidden="true" tabindex="-1"></a>sample_table <span class="op">=</span> get_sample_table(zeller_db)</span>
<span id="cb3-80"><a href="#cb3-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-81"><a href="#cb3-81" aria-hidden="true" tabindex="-1"></a>otu_table.to_csv(<span class="st">'otu_table.csv'</span>)</span>
<span id="cb3-82"><a href="#cb3-82" aria-hidden="true" tabindex="-1"></a>taxa_table.to_csv(<span class="st">'taxa_table.csv'</span>)</span>
<span id="cb3-83"><a href="#cb3-83" aria-hidden="true" tabindex="-1"></a>sample_table.to_csv(<span class="st">'sample_table.csv'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="otu-table" class="level4">
<h4 class="anchored" data-anchor-id="otu-table">OTU table</h4>
<div id="dfb96aea-dfeb-44ac-8661-ff58f2bc1039" class="cell" data-execution_count="125">
<div class="cell-output cell-output-display" data-execution_count="125">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">subjectID</th>
<th data-quarto-table-cell-role="th">FR-001</th>
<th data-quarto-table-cell-role="th">FR-003</th>
<th data-quarto-table-cell-role="th">FR-024</th>
<th data-quarto-table-cell-role="th">FR-025</th>
<th data-quarto-table-cell-role="th">FR-026</th>
<th data-quarto-table-cell-role="th">FR-027</th>
<th data-quarto-table-cell-role="th">FR-030</th>
<th data-quarto-table-cell-role="th">FR-039</th>
<th data-quarto-table-cell-role="th">FR-040</th>
<th data-quarto-table-cell-role="th">FR-051</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">FR-824</th>
<th data-quarto-table-cell-role="th">FR-825</th>
<th data-quarto-table-cell-role="th">FR-826</th>
<th data-quarto-table-cell-role="th">FR-827</th>
<th data-quarto-table-cell-role="th">FR-828</th>
<th data-quarto-table-cell-role="th">FR-829</th>
<th data-quarto-table-cell-role="th">FR-830</th>
<th data-quarto-table-cell-role="th">FR-835</th>
<th data-quarto-table-cell-role="th">FR-901</th>
<th data-quarto-table-cell-role="th">FR-902</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">OTU_0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">OTU_1</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">OTU_2</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">OTU_3</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">OTU_4</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>

<p>5 rows × 156 columns</p>
</div>
</div>
</div>
</div>
</section>
<section id="taxonomy-heirarchy-table" class="level4">
<h4 class="anchored" data-anchor-id="taxonomy-heirarchy-table">Taxonomy heirarchy table</h4>
<div id="2f3151bd-a9aa-4550-9aa0-ee722fae5642" class="cell" data-execution_count="127">
<div class="cell-output cell-output-display" data-execution_count="127">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">kingdom</th>
<th data-quarto-table-cell-role="th">phylum</th>
<th data-quarto-table-cell-role="th">class</th>
<th data-quarto-table-cell-role="th">order</th>
<th data-quarto-table-cell-role="th">family</th>
<th data-quarto-table-cell-role="th">genus</th>
<th data-quarto-table-cell-role="th">species</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">OTU</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">OTU_0</td>
<td>Bacteria</td>
<td>Acidobacteria</td>
<td>Acidobacteriia</td>
<td>Acidobacteriales</td>
<td>Acidobacteriaceae</td>
<td>Granulicella</td>
<td>Granulicella_unclassified</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">OTU_1</td>
<td>Bacteria</td>
<td>Actinobacteria</td>
<td>Actinobacteria</td>
<td>Actinomycetales</td>
<td>Actinomycetaceae</td>
<td>Actinobaculum</td>
<td>Actinobaculum_massiliense</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">OTU_2</td>
<td>Bacteria</td>
<td>Actinobacteria</td>
<td>Actinobacteria</td>
<td>Actinomycetales</td>
<td>Actinomycetaceae</td>
<td>Actinobaculum</td>
<td>Actinobaculum_schaalii</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">OTU_3</td>
<td>Bacteria</td>
<td>Actinobacteria</td>
<td>Actinobacteria</td>
<td>Actinomycetales</td>
<td>Actinomycetaceae</td>
<td>Actinobaculum</td>
<td>Actinobaculum_unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">OTU_4</td>
<td>Bacteria</td>
<td>Actinobacteria</td>
<td>Actinobacteria</td>
<td>Actinomycetales</td>
<td>Actinomycetaceae</td>
<td>Actinomyces</td>
<td>Actinomyces_cardiffensis</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</div>
</section>
<section id="sample-table" class="level4">
<h4 class="anchored" data-anchor-id="sample-table">Sample table</h4>
<div id="94b4a98e-f0f0-4de0-9c9f-ac58fe6fcf44" class="cell" data-execution_count="129">
<div class="cell-output cell-output-display" data-execution_count="129">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">body_site</th>
<th data-quarto-table-cell-role="th">study_condition</th>
<th data-quarto-table-cell-role="th">disease</th>
<th data-quarto-table-cell-role="th">age</th>
<th data-quarto-table-cell-role="th">age_category</th>
<th data-quarto-table-cell-role="th">gender</th>
<th data-quarto-table-cell-role="th">country</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">subjectID</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">FR-001</td>
<td>stool</td>
<td>control</td>
<td>none</td>
<td>52</td>
<td>adult</td>
<td>female</td>
<td>FRA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">FR-003</td>
<td>stool</td>
<td>control</td>
<td>none</td>
<td>70</td>
<td>senior</td>
<td>male</td>
<td>FRA</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">FR-024</td>
<td>stool</td>
<td>control</td>
<td>none</td>
<td>37</td>
<td>adult</td>
<td>female</td>
<td>FRA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">FR-025</td>
<td>stool</td>
<td>control</td>
<td>none</td>
<td>80</td>
<td>senior</td>
<td>female</td>
<td>FRA</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">FR-026</td>
<td>stool</td>
<td>adenoma</td>
<td>adenoma</td>
<td>66</td>
<td>senior</td>
<td>male</td>
<td>FRA</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="visualizing-species-distribution-between-healthy-and-crc-patients" class="level2">
<h2 class="anchored" data-anchor-id="visualizing-species-distribution-between-healthy-and-crc-patients">Visualizing species distribution between healthy and CRC patients</h2>
<p>We will now visualize the distribution of species in the dataset among healthy and CRC patients.</p>
<p>For that we will first aggregate data on family level. The table below shows aggregated abundance at phylum rank.</p>
<div id="9675cbbe-c09e-4014-b691-3d556b9dfef9" class="cell" data-execution_count="3">
<div class="cell-output cell-output-display" data-execution_count="3">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Acidobacteria</th>
<th data-quarto-table-cell-role="th">Actinobacteria</th>
<th data-quarto-table-cell-role="th">Bacteroidetes</th>
<th data-quarto-table-cell-role="th">Candidatus_Saccharibacteria</th>
<th data-quarto-table-cell-role="th">Deferribacteres</th>
<th data-quarto-table-cell-role="th">Deinococcus_Thermus</th>
<th data-quarto-table-cell-role="th">Firmicutes</th>
<th data-quarto-table-cell-role="th">Fusobacteria</th>
<th data-quarto-table-cell-role="th">Proteobacteria</th>
<th data-quarto-table-cell-role="th">Spirochaetes</th>
<th data-quarto-table-cell-role="th">Synergistetes</th>
<th data-quarto-table-cell-role="th">Verrucomicrobia</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">subjectID</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">FR-001</td>
<td>0.0</td>
<td>0.00000</td>
<td>17.89904</td>
<td>0.00000</td>
<td>0.0</td>
<td>0.0</td>
<td>72.93033</td>
<td>0.0</td>
<td>0.61829</td>
<td>0.0</td>
<td>0.0</td>
<td>0.55267</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">FR-003</td>
<td>0.0</td>
<td>0.13650</td>
<td>30.21824</td>
<td>0.01474</td>
<td>0.0</td>
<td>0.0</td>
<td>69.06778</td>
<td>0.0</td>
<td>0.55140</td>
<td>0.0</td>
<td>0.0</td>
<td>0.00000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">FR-024</td>
<td>0.0</td>
<td>13.87554</td>
<td>26.19035</td>
<td>0.00000</td>
<td>0.0</td>
<td>0.0</td>
<td>59.25760</td>
<td>0.0</td>
<td>0.67171</td>
<td>0.0</td>
<td>0.0</td>
<td>0.00000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">FR-025</td>
<td>0.0</td>
<td>9.51098</td>
<td>16.21552</td>
<td>0.00000</td>
<td>0.0</td>
<td>0.0</td>
<td>71.98672</td>
<td>0.0</td>
<td>2.28680</td>
<td>0.0</td>
<td>0.0</td>
<td>0.00000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">FR-026</td>
<td>0.0</td>
<td>0.41172</td>
<td>25.24277</td>
<td>0.00532</td>
<td>0.0</td>
<td>0.0</td>
<td>58.92459</td>
<td>0.0</td>
<td>1.17038</td>
<td>0.0</td>
<td>0.0</td>
<td>0.00000</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</div>
<section id="distribution-at-phylum-level" class="level3">
<h3 class="anchored" data-anchor-id="distribution-at-phylum-level">Distribution at phylum level</h3>
<p>The below figures visualize the distribution of gut microbiome at phylum rank. We can use these plots to select particular phylum for our further analysis. For example, it can clearly seen that <strong>Bacteroides</strong> and <strong>Firnicutes</strong> are two most abundand phylum in the dataset.</p>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>phylum_df <span class="op">=</span> phylum_agg.merge(sample_table, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> plotnine <span class="im">import</span> ggplot, aes, geom_boxplot, theme_minimal, labs, facet_wrap</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>my_pal <span class="op">=</span> {<span class="st">"control"</span>: <span class="st">"g"</span>, <span class="st">"adenoma"</span>: <span class="st">"b"</span>, <span class="st">"CRC"</span>:<span class="st">"r"</span>}</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ph <span class="kw">in</span> phylum_agg.columns:</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    plt.figure()</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    sns.boxplot(x<span class="op">=</span><span class="st">'study_condition'</span>, y<span class="op">=</span>ph, data<span class="op">=</span>phylum_df, palette<span class="op">=</span>my_pal, saturation<span class="op">=</span><span class="fl">0.2</span>, hue<span class="op">=</span><span class="st">'study_condition'</span>)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="st">'</span><span class="sc">{}</span><span class="st"> distribution'</span>.<span class="bu">format</span>(ph))</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-charts" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-charts-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-1.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Acidobacteria
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-2.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Actinobacteria
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-3" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-3.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(c) Bacteroidetes
</figcaption>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-4" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-4.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(d) Candidatus_Saccharibacteria
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-5" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-5.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(e) Deferribacteres
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-6" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-6-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-6.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-6-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(f) Deinococcus_Thermus
</figcaption>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-7" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-7.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(g) Firmicutes
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-8" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-8-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-8.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-8-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(h) Fusobacteria
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-9" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-9.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(i) Proteobacteria
</figcaption>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-10" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-10-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-10.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-10-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(j) Spirochaetes
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-11" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-11-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-11.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-11-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(k) Synergistetes
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts-12" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts-12-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts-output-12.png" class="img-fluid figure-img" data-ref-parent="fig-charts">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts-12-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(l) Verrucomicrobia
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-charts-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Distribution at phylum rank
</figcaption>
</figure>
</div>
<p>Let’s move forward with visualization and plot species level distribution for aforemention phylums.</p>
</section>
<section id="distribution-at-genus-level" class="level3">
<h3 class="anchored" data-anchor-id="distribution-at-genus-level">Distribution at genus level</h3>
<p>At this phase, we filtered out genus with less than <strong>3%</strong> abundance. Then we plotted distribution at genus level for healthy, adenoma and CRC patients. The figures below show distriution of <strong>8</strong> different types of genus in the dataset whose relative abundance satisfied the criteria (&gt;3%).</p>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>filter_phylum <span class="op">=</span> taxa_table.loc[taxa_table[<span class="st">'phylum'</span>].isin([<span class="st">'Bacteroidetes'</span>, <span class="st">'Firmicutes'</span>]),:]</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>genus_fil <span class="op">=</span> aggregate_by_taxonomy(otu_table,filter_phylum,<span class="st">'genus'</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>genus_abund <span class="op">=</span> genus_fil.mean(axis<span class="op">=</span><span class="dv">0</span>).to_dict()</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>genus_list <span class="op">=</span> []</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> genus, value <span class="kw">in</span> genus_abund.items():</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> value <span class="op">&gt;</span> <span class="dv">3</span>:</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>        genus_list.append(genus)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>genus_new <span class="op">=</span> genus_fil[genus_list]</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>genus_fil_df <span class="op">=</span> genus_new.merge(sample_table, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ph <span class="kw">in</span> genus_new.columns:</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    plt.figure()</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    sns.boxplot(x<span class="op">=</span><span class="st">'study_condition'</span>, y<span class="op">=</span>ph, data<span class="op">=</span>genus_fil_df, palette<span class="op">=</span>my_pal, saturation<span class="op">=</span><span class="fl">0.2</span>, hue<span class="op">=</span><span class="st">'study_condition'</span>)</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="st">'</span><span class="sc">{}</span><span class="st"> distribution'</span>.<span class="bu">format</span>(ph))</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-charts2" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-charts2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts2" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts2-1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts2-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts2-output-1.png" class="img-fluid figure-img" data-ref-parent="fig-charts2">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts2-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Bacteroides
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts2" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts2-2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts2-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts2-output-2.png" class="img-fluid figure-img" data-ref-parent="fig-charts2">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts2-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Prevotella
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts2" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts2-3" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts2-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts2-output-3.png" class="img-fluid figure-img" data-ref-parent="fig-charts2">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts2-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(c) Eubacterium
</figcaption>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts2" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts2-4" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts2-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts2-output-4.png" class="img-fluid figure-img" data-ref-parent="fig-charts2">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts2-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(d) Blautia
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts2" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts2-5" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts2-5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts2-output-5.png" class="img-fluid figure-img" data-ref-parent="fig-charts2">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts2-5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(e) Coprococcus
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts2" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts2-6" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts2-6-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts2-output-6.png" class="img-fluid figure-img" data-ref-parent="fig-charts2">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts2-6-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(f) Faecalibacterium
</figcaption>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts2" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts2-7" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts2-7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts2-output-7.png" class="img-fluid figure-img" data-ref-parent="fig-charts2">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts2-7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(g) Ruminococcus
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts2" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts2-8" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts2-8-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts2-output-8.png" class="img-fluid figure-img" data-ref-parent="fig-charts2">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts2-8-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(h) Subdoligranulum
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-charts2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Distribution at genus rank
</figcaption>
</figure>
</div>
</section>
<section id="statistical-significance" class="level3">
<h3 class="anchored" data-anchor-id="statistical-significance">Statistical significance</h3>
<p>Based on our understanding of different taxa at the genus level, we filtered out 8 different types of genus for our statistical analysis. We will use <strong>Kruska Wallis</strong> statistical test for testing the null hypothesis that the <strong>medians of relative abundance of microbial organisms of our selected genus levels are same</strong>.</p>
<p>There are two different genus (Eubacterium and Ruminococcus) whose differences among healthy, adenoma and CRC patients were found statistically significant.</p>
<table class="table">
<thead>
<tr class="header">
<th>Genus</th>
<th>p_value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Bacteroides</td>
<td>0.195</td>
</tr>
<tr class="even">
<td>Prevotella</td>
<td>0.237</td>
</tr>
<tr class="odd">
<td><strong>Eubacterium</strong></td>
<td>0.000</td>
</tr>
<tr class="even">
<td>Blautia</td>
<td>0.479</td>
</tr>
<tr class="odd">
<td>Coprococcus</td>
<td>0.063</td>
</tr>
<tr class="even">
<td>Faecalibacterium</td>
<td>0.131</td>
</tr>
<tr class="odd">
<td><strong>Ruminococcus</strong></td>
<td>0.004</td>
</tr>
<tr class="even">
<td>Subdoligranulum</td>
<td>0.365</td>
</tr>
</tbody>
</table>
</section>
<section id="alpha-diversity" class="level3">
<h3 class="anchored" data-anchor-id="alpha-diversity">Alpha diversity</h3>
<p>We will now investigate whether their are any differences between healthy, adenoma and CRC patients in terms of alpha diversity.</p>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ph <span class="kw">in</span> alpha_diversity_metrics:</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    plt.figure()</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    sns.boxplot(x<span class="op">=</span><span class="st">'study_condition'</span>, y<span class="op">=</span>ph, data<span class="op">=</span>diversity_df, palette<span class="op">=</span>my_pal, saturation<span class="op">=</span><span class="fl">0.2</span>, hue<span class="op">=</span><span class="st">'study_condition'</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="st">'</span><span class="sc">{}</span><span class="st"> metric'</span>.<span class="bu">format</span>(ph))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-charts3" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-charts3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts3" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts3-1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts3-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts3-output-1.png" class="img-fluid figure-img" data-ref-parent="fig-charts3">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts3-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) chao1
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts3" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts3-2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts3-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts3-output-2.png" class="img-fluid figure-img" data-ref-parent="fig-charts3">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts3-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) shannon
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts3" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts3-3" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts3-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts3-output-3.png" class="img-fluid figure-img" data-ref-parent="fig-charts3">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts3-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(c) simpson
</figcaption>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts3" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts3-4" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts3-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts3-output-4.png" class="img-fluid figure-img" data-ref-parent="fig-charts3">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts3-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(d) simpson_e
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts3" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts3-5" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts3-5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts3-output-5.png" class="img-fluid figure-img" data-ref-parent="fig-charts3">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts3-5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(e) fisher_alpha
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts3" style="flex-basis: 33.3%;justify-content: flex-start;">
<div id="fig-charts3-6" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts3-6-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts3-output-6.png" class="img-fluid figure-img" data-ref-parent="fig-charts3">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts3-6-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(f) berger_parker
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-charts3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Alpha diversity measures
</figcaption>
</figure>
</div>
</section>
</section>
<section id="pca-analysis" class="level1">
<h1>PCA analysis</h1>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> StandardScaler</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.decomposition <span class="im">import</span> PCA</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> FunctionTransformer</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.manifold <span class="im">import</span> TSNE</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>genus_scaled <span class="op">=</span> log_transformer.fit_transform(genus_agg)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>pca <span class="op">=</span> PCA(random_state<span class="op">=</span><span class="dv">42</span>,n_components<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>components <span class="op">=</span> pca.fit_transform(genus_scaled)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>c_df <span class="op">=</span> pd.DataFrame(components, columns<span class="op">=</span>[<span class="st">'pca-1'</span>,<span class="st">'pca-2'</span>])</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>c_df.index <span class="op">=</span> sample_table.index</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>comp_df <span class="op">=</span> c_df.merge(sample_table, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(comp_df2,x<span class="op">=</span><span class="st">'pca-1'</span>,y<span class="op">=</span><span class="st">'pca-2'</span>,hue<span class="op">=</span><span class="st">'study_condition'</span>,alpha<span class="op">=</span><span class="fl">.7</span>)</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'PCA'</span>)</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>tsne <span class="op">=</span> TSNE(n_components<span class="op">=</span><span class="dv">2</span>, perplexity<span class="op">=</span><span class="dv">30</span>, random_state<span class="op">=</span><span class="dv">42</span>)</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>tsne_comps <span class="op">=</span> tsne.fit_transform(genus_scaled)</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>t_df <span class="op">=</span> pd.DataFrame(tsne_comps, columns<span class="op">=</span>[<span class="st">'tsne-1'</span>,<span class="st">'tsne-2'</span>])</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>t_df.index <span class="op">=</span> sample_table.index</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>tsne_df <span class="op">=</span> t_df.merge(sample_table, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(tsne_df,x<span class="op">=</span><span class="st">'tsne-1'</span>,y<span class="op">=</span><span class="st">'tsne-2'</span>,hue<span class="op">=</span><span class="st">'study_condition'</span>,alpha<span class="op">=</span><span class="fl">.7</span>)</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'t-SNE'</span>)</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-charts4" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-charts4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts4" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-charts4-1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts4-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts4-output-1.png" class="img-fluid figure-img" data-ref-parent="fig-charts4">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts4-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) PCA with genus abundance
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts4" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-charts4-2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts4-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts4-output-2.png" class="img-fluid figure-img" data-ref-parent="fig-charts4">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts4-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) t-SNE with genus abundance
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-charts4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Dimensionality reduction
</figcaption>
</figure>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> StandardScaler</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.decomposition <span class="im">import</span> PCA</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> FunctionTransformer</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.manifold <span class="im">import</span> TSNE</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>genus_scaled <span class="op">=</span> log_transformer.fit_transform(phylum_agg)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>pca <span class="op">=</span> PCA(random_state<span class="op">=</span><span class="dv">42</span>,n_components<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>components <span class="op">=</span> pca.fit_transform(genus_scaled)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>c_df <span class="op">=</span> pd.DataFrame(components, columns<span class="op">=</span>[<span class="st">'pca-1'</span>,<span class="st">'pca-2'</span>])</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>c_df.index <span class="op">=</span> sample_table.index</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>comp_df <span class="op">=</span> c_df.merge(sample_table, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(comp_df2,x<span class="op">=</span><span class="st">'pca-1'</span>,y<span class="op">=</span><span class="st">'pca-2'</span>,hue<span class="op">=</span><span class="st">'study_condition'</span>,alpha<span class="op">=</span><span class="fl">.7</span>)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'PCA'</span>)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>tsne <span class="op">=</span> TSNE(n_components<span class="op">=</span><span class="dv">2</span>, perplexity<span class="op">=</span><span class="dv">30</span>, random_state<span class="op">=</span><span class="dv">42</span>)</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>tsne_comps <span class="op">=</span> tsne.fit_transform(genus_scaled)</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>t_df <span class="op">=</span> pd.DataFrame(tsne_comps, columns<span class="op">=</span>[<span class="st">'tsne-1'</span>,<span class="st">'tsne-2'</span>])</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>t_df.index <span class="op">=</span> sample_table.index</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>tsne_df <span class="op">=</span> t_df.merge(sample_table, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(tsne_df,x<span class="op">=</span><span class="st">'tsne-1'</span>,y<span class="op">=</span><span class="st">'tsne-2'</span>,hue<span class="op">=</span><span class="st">'study_condition'</span>,alpha<span class="op">=</span><span class="fl">.7</span>)</span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'t-SNE'</span>)</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-charts5" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-charts5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts5" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-charts5-1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts5-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts5-output-1.png" class="img-fluid figure-img" data-ref-parent="fig-charts5">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts5-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) PCA with phylum abundance
</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-charts5" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-charts5-2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-charts5-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="meta_analysis_files/figure-html/fig-charts5-output-2.png" class="img-fluid figure-img" data-ref-parent="fig-charts5">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-charts5-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) t-SNE with phylum abundance
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-charts5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Dimensionality reduction
</figcaption>
</figure>
</div>
</section>
<section id="references" class="level1">
<h1>References</h1>
<p>Arumugam, M., Raes, J., Pelletier, E., Le Paslier, D., Yamada, T., Mende, D. R., … &amp; Bork, P. (2011). Enterotypes of the human gut microbiome. nature, 473(7346), 174-180.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>